[BUG] test_hash_grpby_sum_full_decimal fails

On the UCX cluster, `test_hash_grpby_sum_full_decimal` is failing in a few cases:
```
[2022-08-30T18:29:43.746Z] _ test_hash_grpby_sum_full_decimal[{'spark.rapids.sql.variableFloatAgg.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.castStringToFloat.enabled': 'true'}-[('a', RepeatSeq(Short)), ('b', Decimal(38,-10)), ('c', Decimal(38,-10))]] _
[2022-08-30T18:29:43.746Z] 
[2022-08-30T18:29:43.746Z] data_gen = [('a', RepeatSeq(Short)), ('b', Decimal(38,-10)), ('c', Decimal(38,-10))]
[2022-08-30T18:29:43.746Z] conf = {'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.variableFloatAgg.enabled': 'true'}
[2022-08-30T18:29:43.746Z] 
[2022-08-30T18:29:43.746Z]     @shuffle_test
[2022-08-30T18:29:43.746Z]     @approximate_float
[2022-08-30T18:29:43.746Z]     @ignore_order
[2022-08-30T18:29:43.746Z]     @incompat
[2022-08-30T18:29:43.746Z]     @pytest.mark.parametrize('data_gen', _init_list_full_decimal, ids=idfn)
[2022-08-30T18:29:43.746Z]     @pytest.mark.parametrize('conf', get_params(_confs, params_markers_for_confs), ids=idfn)
[2022-08-30T18:29:43.746Z]     def test_hash_grpby_sum_full_decimal(data_gen, conf):
[2022-08-30T18:29:43.746Z] >       assert_gpu_and_cpu_are_equal_collect(
[2022-08-30T18:29:43.747Z]             lambda spark: gen_df(spark, data_gen, length=100).groupby('a').agg(f.sum('b')),
[2022-08-30T18:29:43.747Z]             conf = conf)
[2022-08-30T18:29:43.747Z] 
[2022-08-30T18:29:43.747Z] ../../src/main/python/hash_aggregate_test.py:379: 
[2022-08-30T18:29:43.747Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:508: in assert_gpu_and_cpu_are_equal_collect
[2022-08-30T18:29:43.747Z]     _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:439: in _assert_gpu_and_cpu_are_equal
[2022-08-30T18:29:43.747Z]     assert_equal(from_cpu, from_gpu)
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:106: in assert_equal
[2022-08-30T18:29:43.747Z]     _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:42: in _assert_equal
[2022-08-30T18:29:43.747Z]     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:35: in _assert_equal
[2022-08-30T18:29:43.747Z]     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2022-08-30T18:29:43.747Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2022-08-30T18:29:43.747Z] 
[2022-08-30T18:29:43.747Z] cpu = None, gpu = Decimal('-1.0069950567342662336581895246262384447E+47')
[2022-08-30T18:29:43.747Z] float_check = <function get_float_check.<locals>.<lambda> at 0x7f7b0924f430>
[2022-08-30T18:29:43.747Z] path = [47, 'sum(b)']
[2022-08-30T18:29:43.747Z] 
[2022-08-30T18:29:43.747Z]     def _assert_equal(cpu, gpu, float_check, path):
[2022-08-30T18:29:43.747Z]         t = type(cpu)
[2022-08-30T18:29:43.747Z]         if (t is Row):
[2022-08-30T18:29:43.747Z]             assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.747Z]             if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
[2022-08-30T18:29:43.747Z]                 assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
[2022-08-30T18:29:43.747Z]                 for field in cpu.__fields__:
[2022-08-30T18:29:43.747Z]                     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2022-08-30T18:29:43.747Z]             else:
[2022-08-30T18:29:43.747Z]                 for index in range(len(cpu)):
[2022-08-30T18:29:43.747Z]                     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.747Z]         elif (t is list):
[2022-08-30T18:29:43.747Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.747Z]             for index in range(len(cpu)):
[2022-08-30T18:29:43.747Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.747Z]         elif (t is tuple):
[2022-08-30T18:29:43.747Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.747Z]             for index in range(len(cpu)):
[2022-08-30T18:29:43.747Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.747Z]         elif (t is pytypes.GeneratorType):
[2022-08-30T18:29:43.747Z]             index = 0
[2022-08-30T18:29:43.747Z]             # generator has no zip :( so we have to do this the hard way
[2022-08-30T18:29:43.747Z]             done = False
[2022-08-30T18:29:43.747Z]             while not done:
[2022-08-30T18:29:43.747Z]                 sub_cpu = None
[2022-08-30T18:29:43.747Z]                 sub_gpu = None
[2022-08-30T18:29:43.747Z]                 try:
[2022-08-30T18:29:43.747Z]                     sub_cpu = next(cpu)
[2022-08-30T18:29:43.747Z]                 except StopIteration:
[2022-08-30T18:29:43.747Z]                     done = True
[2022-08-30T18:29:43.747Z]     
[2022-08-30T18:29:43.747Z]                 try:
[2022-08-30T18:29:43.747Z]                     sub_gpu = next(gpu)
[2022-08-30T18:29:43.747Z]                 except StopIteration:
[2022-08-30T18:29:43.747Z]                     done = True
[2022-08-30T18:29:43.747Z]     
[2022-08-30T18:29:43.747Z]                 if done:
[2022-08-30T18:29:43.747Z]                     assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
[2022-08-30T18:29:43.747Z]                 else:
[2022-08-30T18:29:43.747Z]                     _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])
[2022-08-30T18:29:43.747Z]     
[2022-08-30T18:29:43.747Z]                 index = index + 1
[2022-08-30T18:29:43.747Z]         elif (t is dict):
[2022-08-30T18:29:43.747Z]             # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
[2022-08-30T18:29:43.747Z]             # so sort the items to do our best with ignoring the order of dicts
[2022-08-30T18:29:43.747Z]             cpu_items = list(cpu.items()).sort(key=_RowCmp)
[2022-08-30T18:29:43.747Z]             gpu_items = list(gpu.items()).sort(key=_RowCmp)
[2022-08-30T18:29:43.747Z]             _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
[2022-08-30T18:29:43.747Z]         elif (t is int):
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif (t is float):
[2022-08-30T18:29:43.747Z]             if (math.isnan(cpu)):
[2022-08-30T18:29:43.747Z]                 assert math.isnan(gpu), "GPU and CPU float values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]             else:
[2022-08-30T18:29:43.747Z]                 assert float_check(cpu, gpu), "GPU and CPU float values are different {}".format(path)
[2022-08-30T18:29:43.747Z]         elif isinstance(cpu, str):
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU string values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif isinstance(cpu, datetime):
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU timestamp values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif isinstance(cpu, date):
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU date values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif isinstance(cpu, bool):
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU boolean values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif isinstance(cpu, Decimal):
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU decimal values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif isinstance(cpu, bytearray):
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU bytearray values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif isinstance(cpu, timedelta):
[2022-08-30T18:29:43.747Z]             # Used by interval type DayTimeInterval for Pyspark 3.3.0+
[2022-08-30T18:29:43.747Z]             assert cpu == gpu, "GPU and CPU timedelta values are different at {}".format(path)
[2022-08-30T18:29:43.747Z]         elif (cpu == None):
[2022-08-30T18:29:43.747Z] >           assert cpu == gpu, "GPU and CPU are not both null at {}".format(path)
[2022-08-30T18:29:43.747Z] E           AssertionError: GPU and CPU are not both null at [47, 'sum(b)']
[2022-08-30T18:29:43.747Z] 
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:99: AssertionError
[2022-08-30T18:29:43.747Z] ----------------------------- Captured stdout call -----------------------------
[2022-08-30T18:29:43.747Z] ### CPU RUN ###
[2022-08-30T18:29:43.747Z] ### GPU RUN ###
[2022-08-30T18:29:43.747Z] ### COLLECT: GPU TOOK 0.2465381622314453 CPU TOOK 0.2997779846191406 ###
[2022-08-30T18:29:43.747Z] CPU OUTPUT: [Row(a=-32619, sum(b)=None), Row(a=-30928, sum(b)=Decimal('4.3613799014647028251830949477242969843E+47')), Row(a=-28447, sum(b)=Decimal('-8.3725569466846329443175710034175061750E+47')), Row(a=-27704, sum(b)=Decimal('8.3673968153546004357978039265761081121E+47')), Row(a=-24513, sum(b)=None), Row(a=-23103, sum(b)=Decimal('-7.3881712366500820608114571040007436072E+47')), Row(a=-20995, sum(b)=Decimal('-4.3285081513864731207507460740810091460E+47')), Row(a=-20339, sum(b)=Decimal('7.8026661452699550461353941661132082832E+47')), Row(a=-19569, sum(b)=Decimal('3.7944782979790961693390352992571616855E+47')), Row(a=-17649, sum(b)=Decimal('-6.3811224080804710763061284537997550196E+47')), Row(a=-15700, sum(b)=Decimal('-1.7835153556073193480854495404826110615E+47')), Row(a=-14091, sum(b)=Decimal('-7.7578447058772832613753473514625789601E+47')), Row(a=-13506, sum(b)=Decimal('-1.1865363805792297674977806356804951307E+47')), Row(a=-7945, sum(b)=Decimal('2.1642604475137131231503773004203997695E+47')), Row(a=-6134, sum(b)=Decimal('6.5182502818993305090352473765013963143E+47')), Row(a=-5967, sum(b)=None), Row(a=-4137, sum(b)=Decimal('3.3111966566756745404264093656831713319E+47')), Row(a=-3709, sum(b)=Decimal('2.1867479065020489288063726466395621356E+47')), Row(a=-1901, sum(b)=Decimal('4.729335124079427289918337653286636193E+46')), Row(a=-1035, sum(b)=Decimal('-1.2777901277508984890515648228508830240E+47')), Row(a=-913, sum(b)=Decimal('-8.163431550632523582283561564508231025E+46')), Row(a=1, sum(b)=Decimal('6.1905973175579656937274349688102686361E+47')), Row(a=66, sum(b)=None), Row(a=1318, sum(b)=Decimal('6.3636197454468312157141065752120354598E+47')), Row(a=1375, sum(b)=Decimal('-3.7106304908383885223095895931328039293E+47')), Row(a=3359, sum(b)=Decimal('-5.9682615996904006072299884545466058266E+47')), Row(a=4173, sum(b)=Decimal('7.53217583143941892444266253113610777E+45')), Row(a=4935, sum(b)=Decimal('-3.525048965644564623348132531719032994E+46')), Row(a=5385, sum(b)=Decimal('-7.52015698258531592517770267898834856E+45')), Row(a=6987, sum(b)=Decimal('-5.5449721492552717064150016930487718249E+47')), Row(a=8676, sum(b)=Decimal('3.4114832387670067691338977572134299574E+47')), Row(a=9182, sum(b)=Decimal('5.4628090452150896307406820551619391806E+47')), Row(a=9857, sum(b)=Decimal('2.688551847197926787241363712285667881E+46')), Row(a=10511, sum(b)=Decimal('2.3829901252007883888494414498035548199E+47')), Row(a=10846, sum(b)=None), Row(a=10896, sum(b)=Decimal('-3.6635001565727457654572234976512213543E+47')), Row(a=11176, sum(b)=Decimal('-4.9903919604467173684789641371226472273E+47')), Row(a=14072, sum(b)=Decimal('-7.5741626831426962049074365723075317754E+47')), Row(a=17681, sum(b)=Decimal('2.5540633446776025449687473521701111467E+47')), Row(a=18519, sum(b)=Decimal('-4.3213708809092373414707178568975625170E+47')), Row(a=19506, sum(b)=Decimal('-1.1531723562438250311691614317433377622E+47')), Row(a=22206, sum(b)=None), Row(a=22357, sum(b)=Decimal('-7.4475523813202712337620849352375571012E+47')), Row(a=25948, sum(b)=Decimal('-7.0518383647729405524064484412608892422E+47')), Row(a=26288, sum(b)=Decimal('-4.4675581455047983491819071764736994108E+47')), Row(a=30923, sum(b)=Decimal('-9.6375396703039112292881835675701043809E+47')), Row(a=31363, sum(b)=None), Row(a=32767, sum(b)=None)]
[2022-08-30T18:29:43.747Z] GPU OUTPUT: [Row(a=-32619, sum(b)=None), Row(a=-30928, sum(b)=Decimal('4.3613799014647028251830949477242969843E+47')), Row(a=-28447, sum(b)=Decimal('-8.3725569466846329443175710034175061750E+47')), Row(a=-27704, sum(b)=Decimal('8.3673968153546004357978039265761081121E+47')), Row(a=-24513, sum(b)=None), Row(a=-23103, sum(b)=Decimal('-7.3881712366500820608114571040007436072E+47')), Row(a=-20995, sum(b)=Decimal('-4.3285081513864731207507460740810091460E+47')), Row(a=-20339, sum(b)=Decimal('7.8026661452699550461353941661132082832E+47')), Row(a=-19569, sum(b)=Decimal('3.7944782979790961693390352992571616855E+47')), Row(a=-17649, sum(b)=Decimal('-6.3811224080804710763061284537997550196E+47')), Row(a=-15700, sum(b)=Decimal('-1.7835153556073193480854495404826110615E+47')), Row(a=-14091, sum(b)=Decimal('-7.7578447058772832613753473514625789601E+47')), Row(a=-13506, sum(b)=Decimal('-1.1865363805792297674977806356804951307E+47')), Row(a=-7945, sum(b)=Decimal('2.1642604475137131231503773004203997695E+47')), Row(a=-6134, sum(b)=Decimal('6.5182502818993305090352473765013963143E+47')), Row(a=-5967, sum(b)=None), Row(a=-4137, sum(b)=Decimal('3.3111966566756745404264093656831713319E+47')), Row(a=-3709, sum(b)=Decimal('2.1867479065020489288063726466395621356E+47')), Row(a=-1901, sum(b)=Decimal('4.729335124079427289918337653286636193E+46')), Row(a=-1035, sum(b)=Decimal('-1.2777901277508984890515648228508830240E+47')), Row(a=-913, sum(b)=Decimal('-8.163431550632523582283561564508231025E+46')), Row(a=1, sum(b)=Decimal('6.1905973175579656937274349688102686361E+47')), Row(a=66, sum(b)=None), Row(a=1318, sum(b)=Decimal('6.3636197454468312157141065752120354598E+47')), Row(a=1375, sum(b)=Decimal('-3.7106304908383885223095895931328039293E+47')), Row(a=3359, sum(b)=Decimal('-5.9682615996904006072299884545466058266E+47')), Row(a=4173, sum(b)=Decimal('7.53217583143941892444266253113610777E+45')), Row(a=4935, sum(b)=Decimal('-3.525048965644564623348132531719032994E+46')), Row(a=5385, sum(b)=Decimal('-7.52015698258531592517770267898834856E+45')), Row(a=6987, sum(b)=Decimal('-5.5449721492552717064150016930487718249E+47')), Row(a=8676, sum(b)=Decimal('3.4114832387670067691338977572134299574E+47')), Row(a=9182, sum(b)=Decimal('5.4628090452150896307406820551619391806E+47')), Row(a=9857, sum(b)=Decimal('2.688551847197926787241363712285667881E+46')), Row(a=10511, sum(b)=Decimal('2.3829901252007883888494414498035548199E+47')), Row(a=10846, sum(b)=None), Row(a=10896, sum(b)=Decimal('-3.6635001565727457654572234976512213543E+47')), Row(a=11176, sum(b)=Decimal('-4.9903919604467173684789641371226472273E+47')), Row(a=14072, sum(b)=Decimal('-7.5741626831426962049074365723075317754E+47')), Row(a=17681, sum(b)=Decimal('2.5540633446776025449687473521701111467E+47')), Row(a=18519, sum(b)=Decimal('-4.3213708809092373414707178568975625170E+47')), Row(a=19506, sum(b)=Decimal('-1.1531723562438250311691614317433377622E+47')), Row(a=22206, sum(b)=None), Row(a=22357, sum(b)=Decimal('-7.4475523813202712337620849352375571012E+47')), Row(a=25948, sum(b)=Decimal('-7.0518383647729405524064484412608892422E+47')), Row(a=26288, sum(b)=Decimal('-4.4675581455047983491819071764736994108E+47')), Row(a=30923, sum(b)=Decimal('-9.6375396703039112292881835675701043809E+47')), Row(a=31363, sum(b)=None), Row(a=32767, sum(b)=Decimal('-1.0069950567342662336581895246262384447E+47'))]
[2022-08-30T18:29:43.747Z] _ test_hash_grpby_sum_full_decimal[{'spark.rapids.sql.variableFloatAgg.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hashAgg.replaceMode': 'final'}-[('a', RepeatSeq(Short)), ('b', Decimal(38,0)), ('c', Decimal(38,0))]] _
[2022-08-30T18:29:43.747Z] 
[2022-08-30T18:29:43.747Z] data_gen = [('a', RepeatSeq(Short)), ('b', Decimal(38,0)), ('c', Decimal(38,0))]
[2022-08-30T18:29:43.747Z] conf = {'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.hashAgg.replaceMode': 'final', 'spark.rapids.sql.variableFloatAgg.enabled': 'true'}
[2022-08-30T18:29:43.747Z] 
[2022-08-30T18:29:43.747Z]     @shuffle_test
[2022-08-30T18:29:43.747Z]     @approximate_float
[2022-08-30T18:29:43.747Z]     @ignore_order
[2022-08-30T18:29:43.747Z]     @incompat
[2022-08-30T18:29:43.747Z]     @pytest.mark.parametrize('data_gen', _init_list_full_decimal, ids=idfn)
[2022-08-30T18:29:43.747Z]     @pytest.mark.parametrize('conf', get_params(_confs, params_markers_for_confs), ids=idfn)
[2022-08-30T18:29:43.747Z]     def test_hash_grpby_sum_full_decimal(data_gen, conf):
[2022-08-30T18:29:43.747Z] >       assert_gpu_and_cpu_are_equal_collect(
[2022-08-30T18:29:43.747Z]             lambda spark: gen_df(spark, data_gen, length=100).groupby('a').agg(f.sum('b')),
[2022-08-30T18:29:43.747Z]             conf = conf)
[2022-08-30T18:29:43.747Z] 
[2022-08-30T18:29:43.747Z] ../../src/main/python/hash_aggregate_test.py:379: 
[2022-08-30T18:29:43.747Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:508: in assert_gpu_and_cpu_are_equal_collect
[2022-08-30T18:29:43.747Z]     _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
[2022-08-30T18:29:43.747Z] ../../src/main/python/asserts.py:439: in _assert_gpu_and_cpu_are_equal
[2022-08-30T18:29:43.747Z]     assert_equal(from_cpu, from_gpu)
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:106: in assert_equal
[2022-08-30T18:29:43.748Z]     _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:42: in _assert_equal
[2022-08-30T18:29:43.748Z]     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:35: in _assert_equal
[2022-08-30T18:29:43.748Z]     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2022-08-30T18:29:43.748Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] cpu = None, gpu = Decimal('-10069950567342662336581895246262384447')
[2022-08-30T18:29:43.748Z] float_check = <function get_float_check.<locals>.<lambda> at 0x7f7b088528b0>
[2022-08-30T18:29:43.748Z] path = [47, 'sum(b)']
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z]     def _assert_equal(cpu, gpu, float_check, path):
[2022-08-30T18:29:43.748Z]         t = type(cpu)
[2022-08-30T18:29:43.748Z]         if (t is Row):
[2022-08-30T18:29:43.748Z]             assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.748Z]             if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
[2022-08-30T18:29:43.748Z]                 assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
[2022-08-30T18:29:43.748Z]                 for field in cpu.__fields__:
[2022-08-30T18:29:43.748Z]                     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2022-08-30T18:29:43.748Z]             else:
[2022-08-30T18:29:43.748Z]                 for index in range(len(cpu)):
[2022-08-30T18:29:43.748Z]                     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z]         elif (t is list):
[2022-08-30T18:29:43.748Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.748Z]             for index in range(len(cpu)):
[2022-08-30T18:29:43.748Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z]         elif (t is tuple):
[2022-08-30T18:29:43.748Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.748Z]             for index in range(len(cpu)):
[2022-08-30T18:29:43.748Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z]         elif (t is pytypes.GeneratorType):
[2022-08-30T18:29:43.748Z]             index = 0
[2022-08-30T18:29:43.748Z]             # generator has no zip :( so we have to do this the hard way
[2022-08-30T18:29:43.748Z]             done = False
[2022-08-30T18:29:43.748Z]             while not done:
[2022-08-30T18:29:43.748Z]                 sub_cpu = None
[2022-08-30T18:29:43.748Z]                 sub_gpu = None
[2022-08-30T18:29:43.748Z]                 try:
[2022-08-30T18:29:43.748Z]                     sub_cpu = next(cpu)
[2022-08-30T18:29:43.748Z]                 except StopIteration:
[2022-08-30T18:29:43.748Z]                     done = True
[2022-08-30T18:29:43.748Z]     
[2022-08-30T18:29:43.748Z]                 try:
[2022-08-30T18:29:43.748Z]                     sub_gpu = next(gpu)
[2022-08-30T18:29:43.748Z]                 except StopIteration:
[2022-08-30T18:29:43.748Z]                     done = True
[2022-08-30T18:29:43.748Z]     
[2022-08-30T18:29:43.748Z]                 if done:
[2022-08-30T18:29:43.748Z]                     assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
[2022-08-30T18:29:43.748Z]                 else:
[2022-08-30T18:29:43.748Z]                     _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])
[2022-08-30T18:29:43.748Z]     
[2022-08-30T18:29:43.748Z]                 index = index + 1
[2022-08-30T18:29:43.748Z]         elif (t is dict):
[2022-08-30T18:29:43.748Z]             # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
[2022-08-30T18:29:43.748Z]             # so sort the items to do our best with ignoring the order of dicts
[2022-08-30T18:29:43.748Z]             cpu_items = list(cpu.items()).sort(key=_RowCmp)
[2022-08-30T18:29:43.748Z]             gpu_items = list(gpu.items()).sort(key=_RowCmp)
[2022-08-30T18:29:43.748Z]             _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
[2022-08-30T18:29:43.748Z]         elif (t is int):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif (t is float):
[2022-08-30T18:29:43.748Z]             if (math.isnan(cpu)):
[2022-08-30T18:29:43.748Z]                 assert math.isnan(gpu), "GPU and CPU float values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]             else:
[2022-08-30T18:29:43.748Z]                 assert float_check(cpu, gpu), "GPU and CPU float values are different {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, str):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU string values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, datetime):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU timestamp values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, date):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU date values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, bool):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU boolean values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, Decimal):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU decimal values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, bytearray):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU bytearray values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, timedelta):
[2022-08-30T18:29:43.748Z]             # Used by interval type DayTimeInterval for Pyspark 3.3.0+
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU timedelta values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif (cpu == None):
[2022-08-30T18:29:43.748Z] >           assert cpu == gpu, "GPU and CPU are not both null at {}".format(path)
[2022-08-30T18:29:43.748Z] E           AssertionError: GPU and CPU are not both null at [47, 'sum(b)']
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:99: AssertionError
[2022-08-30T18:29:43.748Z] ----------------------------- Captured stdout call -----------------------------
[2022-08-30T18:29:43.748Z] ### CPU RUN ###
[2022-08-30T18:29:43.748Z] ### GPU RUN ###
[2022-08-30T18:29:43.748Z] ### COLLECT: GPU TOOK 0.27361202239990234 CPU TOOK 0.264432430267334 ###
[2022-08-30T18:29:43.748Z] CPU OUTPUT: [Row(a=-32619, sum(b)=None), Row(a=-30928, sum(b)=Decimal('43613799014647028251830949477242969843')), Row(a=-28447, sum(b)=Decimal('-83725569466846329443175710034175061750')), Row(a=-27704, sum(b)=Decimal('83673968153546004357978039265761081121')), Row(a=-24513, sum(b)=None), Row(a=-23103, sum(b)=Decimal('-73881712366500820608114571040007436072')), Row(a=-20995, sum(b)=Decimal('-43285081513864731207507460740810091460')), Row(a=-20339, sum(b)=Decimal('78026661452699550461353941661132082832')), Row(a=-19569, sum(b)=Decimal('37944782979790961693390352992571616855')), Row(a=-17649, sum(b)=Decimal('-63811224080804710763061284537997550196')), Row(a=-15700, sum(b)=Decimal('-17835153556073193480854495404826110615')), Row(a=-14091, sum(b)=Decimal('-77578447058772832613753473514625789601')), Row(a=-13506, sum(b)=Decimal('-11865363805792297674977806356804951307')), Row(a=-7945, sum(b)=Decimal('21642604475137131231503773004203997695')), Row(a=-6134, sum(b)=Decimal('65182502818993305090352473765013963143')), Row(a=-5967, sum(b)=None), Row(a=-4137, sum(b)=Decimal('33111966566756745404264093656831713319')), Row(a=-3709, sum(b)=Decimal('21867479065020489288063726466395621356')), Row(a=-1901, sum(b)=Decimal('4729335124079427289918337653286636193')), Row(a=-1035, sum(b)=Decimal('-12777901277508984890515648228508830240')), Row(a=-913, sum(b)=Decimal('-8163431550632523582283561564508231025')), Row(a=1, sum(b)=Decimal('61905973175579656937274349688102686361')), Row(a=66, sum(b)=None), Row(a=1318, sum(b)=Decimal('63636197454468312157141065752120354598')), Row(a=1375, sum(b)=Decimal('-37106304908383885223095895931328039293')), Row(a=3359, sum(b)=Decimal('-59682615996904006072299884545466058266')), Row(a=4173, sum(b)=Decimal('753217583143941892444266253113610777')), Row(a=4935, sum(b)=Decimal('-3525048965644564623348132531719032994')), Row(a=5385, sum(b)=Decimal('-752015698258531592517770267898834856')), Row(a=6987, sum(b)=Decimal('-55449721492552717064150016930487718249')), Row(a=8676, sum(b)=Decimal('34114832387670067691338977572134299574')), Row(a=9182, sum(b)=Decimal('54628090452150896307406820551619391806')), Row(a=9857, sum(b)=Decimal('2688551847197926787241363712285667881')), Row(a=10511, sum(b)=Decimal('23829901252007883888494414498035548199')), Row(a=10846, sum(b)=None), Row(a=10896, sum(b)=Decimal('-36635001565727457654572234976512213543')), Row(a=11176, sum(b)=Decimal('-49903919604467173684789641371226472273')), Row(a=14072, sum(b)=Decimal('-75741626831426962049074365723075317754')), Row(a=17681, sum(b)=Decimal('25540633446776025449687473521701111467')), Row(a=18519, sum(b)=Decimal('-43213708809092373414707178568975625170')), Row(a=19506, sum(b)=Decimal('-11531723562438250311691614317433377622')), Row(a=22206, sum(b)=None), Row(a=22357, sum(b)=Decimal('-74475523813202712337620849352375571012')), Row(a=25948, sum(b)=Decimal('-70518383647729405524064484412608892422')), Row(a=26288, sum(b)=Decimal('-44675581455047983491819071764736994108')), Row(a=30923, sum(b)=Decimal('-96375396703039112292881835675701043809')), Row(a=31363, sum(b)=None), Row(a=32767, sum(b)=None)]
[2022-08-30T18:29:43.748Z] GPU OUTPUT: [Row(a=-32619, sum(b)=None), Row(a=-30928, sum(b)=Decimal('43613799014647028251830949477242969843')), Row(a=-28447, sum(b)=Decimal('-83725569466846329443175710034175061750')), Row(a=-27704, sum(b)=Decimal('83673968153546004357978039265761081121')), Row(a=-24513, sum(b)=None), Row(a=-23103, sum(b)=Decimal('-73881712366500820608114571040007436072')), Row(a=-20995, sum(b)=Decimal('-43285081513864731207507460740810091460')), Row(a=-20339, sum(b)=Decimal('78026661452699550461353941661132082832')), Row(a=-19569, sum(b)=Decimal('37944782979790961693390352992571616855')), Row(a=-17649, sum(b)=Decimal('-63811224080804710763061284537997550196')), Row(a=-15700, sum(b)=Decimal('-17835153556073193480854495404826110615')), Row(a=-14091, sum(b)=Decimal('-77578447058772832613753473514625789601')), Row(a=-13506, sum(b)=Decimal('-11865363805792297674977806356804951307')), Row(a=-7945, sum(b)=Decimal('21642604475137131231503773004203997695')), Row(a=-6134, sum(b)=Decimal('65182502818993305090352473765013963143')), Row(a=-5967, sum(b)=None), Row(a=-4137, sum(b)=Decimal('33111966566756745404264093656831713319')), Row(a=-3709, sum(b)=Decimal('21867479065020489288063726466395621356')), Row(a=-1901, sum(b)=Decimal('4729335124079427289918337653286636193')), Row(a=-1035, sum(b)=Decimal('-12777901277508984890515648228508830240')), Row(a=-913, sum(b)=Decimal('-8163431550632523582283561564508231025')), Row(a=1, sum(b)=Decimal('61905973175579656937274349688102686361')), Row(a=66, sum(b)=None), Row(a=1318, sum(b)=Decimal('63636197454468312157141065752120354598')), Row(a=1375, sum(b)=Decimal('-37106304908383885223095895931328039293')), Row(a=3359, sum(b)=Decimal('-59682615996904006072299884545466058266')), Row(a=4173, sum(b)=Decimal('753217583143941892444266253113610777')), Row(a=4935, sum(b)=Decimal('-3525048965644564623348132531719032994')), Row(a=5385, sum(b)=Decimal('-752015698258531592517770267898834856')), Row(a=6987, sum(b)=Decimal('-55449721492552717064150016930487718249')), Row(a=8676, sum(b)=Decimal('34114832387670067691338977572134299574')), Row(a=9182, sum(b)=Decimal('54628090452150896307406820551619391806')), Row(a=9857, sum(b)=Decimal('2688551847197926787241363712285667881')), Row(a=10511, sum(b)=Decimal('23829901252007883888494414498035548199')), Row(a=10846, sum(b)=None), Row(a=10896, sum(b)=Decimal('-36635001565727457654572234976512213543')), Row(a=11176, sum(b)=Decimal('-49903919604467173684789641371226472273')), Row(a=14072, sum(b)=Decimal('-75741626831426962049074365723075317754')), Row(a=17681, sum(b)=Decimal('25540633446776025449687473521701111467')), Row(a=18519, sum(b)=Decimal('-43213708809092373414707178568975625170')), Row(a=19506, sum(b)=Decimal('-11531723562438250311691614317433377622')), Row(a=22206, sum(b)=None), Row(a=22357, sum(b)=Decimal('-74475523813202712337620849352375571012')), Row(a=25948, sum(b)=Decimal('-70518383647729405524064484412608892422')), Row(a=26288, sum(b)=Decimal('-44675581455047983491819071764736994108')), Row(a=30923, sum(b)=Decimal('-96375396703039112292881835675701043809')), Row(a=31363, sum(b)=None), Row(a=32767, sum(b)=Decimal('-10069950567342662336581895246262384447'))]
[2022-08-30T18:29:43.748Z] _ test_hash_grpby_sum_full_decimal[{'spark.rapids.sql.variableFloatAgg.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hashAgg.replaceMode': 'partial'}-[('a', RepeatSeq(Short)), ('b', Decimal(38,0)), ('c', Decimal(38,0))]] _
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] data_gen = [('a', RepeatSeq(Short)), ('b', Decimal(38,0)), ('c', Decimal(38,0))]
[2022-08-30T18:29:43.748Z] conf = {'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.hashAgg.replaceMode': 'partial', 'spark.rapids.sql.variableFloatAgg.enabled': 'true'}
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z]     @shuffle_test
[2022-08-30T18:29:43.748Z]     @approximate_float
[2022-08-30T18:29:43.748Z]     @ignore_order
[2022-08-30T18:29:43.748Z]     @incompat
[2022-08-30T18:29:43.748Z]     @pytest.mark.parametrize('data_gen', _init_list_full_decimal, ids=idfn)
[2022-08-30T18:29:43.748Z]     @pytest.mark.parametrize('conf', get_params(_confs, params_markers_for_confs), ids=idfn)
[2022-08-30T18:29:43.748Z]     def test_hash_grpby_sum_full_decimal(data_gen, conf):
[2022-08-30T18:29:43.748Z] >       assert_gpu_and_cpu_are_equal_collect(
[2022-08-30T18:29:43.748Z]             lambda spark: gen_df(spark, data_gen, length=100).groupby('a').agg(f.sum('b')),
[2022-08-30T18:29:43.748Z]             conf = conf)
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] ../../src/main/python/hash_aggregate_test.py:379: 
[2022-08-30T18:29:43.748Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:508: in assert_gpu_and_cpu_are_equal_collect
[2022-08-30T18:29:43.748Z]     _assert_gpu_and_cpu_are_equal(func, 'COLLECT', conf=conf, is_cpu_first=is_cpu_first)
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:439: in _assert_gpu_and_cpu_are_equal
[2022-08-30T18:29:43.748Z]     assert_equal(from_cpu, from_gpu)
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:106: in assert_equal
[2022-08-30T18:29:43.748Z]     _assert_equal(cpu, gpu, float_check=get_float_check(), path=[])
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:42: in _assert_equal
[2022-08-30T18:29:43.748Z]     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:35: in _assert_equal
[2022-08-30T18:29:43.748Z]     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2022-08-30T18:29:43.748Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] cpu = Decimal('-10069950567342662336581895246262384447'), gpu = None
[2022-08-30T18:29:43.748Z] float_check = <function get_float_check.<locals>.<lambda> at 0x7f7b07a25310>
[2022-08-30T18:29:43.748Z] path = [47, 'sum(b)']
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z]     def _assert_equal(cpu, gpu, float_check, path):
[2022-08-30T18:29:43.748Z]         t = type(cpu)
[2022-08-30T18:29:43.748Z]         if (t is Row):
[2022-08-30T18:29:43.748Z]             assert len(cpu) == len(gpu), "CPU and GPU row have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.748Z]             if hasattr(cpu, "__fields__") and hasattr(gpu, "__fields__"):
[2022-08-30T18:29:43.748Z]                 assert cpu.__fields__ == gpu.__fields__, "CPU and GPU row have different fields at {} CPU: {} GPU: {}".format(path, cpu.__fields__, gpu.__fields__)
[2022-08-30T18:29:43.748Z]                 for field in cpu.__fields__:
[2022-08-30T18:29:43.748Z]                     _assert_equal(cpu[field], gpu[field], float_check, path + [field])
[2022-08-30T18:29:43.748Z]             else:
[2022-08-30T18:29:43.748Z]                 for index in range(len(cpu)):
[2022-08-30T18:29:43.748Z]                     _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z]         elif (t is list):
[2022-08-30T18:29:43.748Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.748Z]             for index in range(len(cpu)):
[2022-08-30T18:29:43.748Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z]         elif (t is tuple):
[2022-08-30T18:29:43.748Z]             assert len(cpu) == len(gpu), "CPU and GPU list have different lengths at {} CPU: {} GPU: {}".format(path, len(cpu), len(gpu))
[2022-08-30T18:29:43.748Z]             for index in range(len(cpu)):
[2022-08-30T18:29:43.748Z]                 _assert_equal(cpu[index], gpu[index], float_check, path + [index])
[2022-08-30T18:29:43.748Z]         elif (t is pytypes.GeneratorType):
[2022-08-30T18:29:43.748Z]             index = 0
[2022-08-30T18:29:43.748Z]             # generator has no zip :( so we have to do this the hard way
[2022-08-30T18:29:43.748Z]             done = False
[2022-08-30T18:29:43.748Z]             while not done:
[2022-08-30T18:29:43.748Z]                 sub_cpu = None
[2022-08-30T18:29:43.748Z]                 sub_gpu = None
[2022-08-30T18:29:43.748Z]                 try:
[2022-08-30T18:29:43.748Z]                     sub_cpu = next(cpu)
[2022-08-30T18:29:43.748Z]                 except StopIteration:
[2022-08-30T18:29:43.748Z]                     done = True
[2022-08-30T18:29:43.748Z]     
[2022-08-30T18:29:43.748Z]                 try:
[2022-08-30T18:29:43.748Z]                     sub_gpu = next(gpu)
[2022-08-30T18:29:43.748Z]                 except StopIteration:
[2022-08-30T18:29:43.748Z]                     done = True
[2022-08-30T18:29:43.748Z]     
[2022-08-30T18:29:43.748Z]                 if done:
[2022-08-30T18:29:43.748Z]                     assert sub_cpu == sub_gpu and sub_cpu == None, "CPU and GPU generators have different lengths at {}".format(path)
[2022-08-30T18:29:43.748Z]                 else:
[2022-08-30T18:29:43.748Z]                     _assert_equal(sub_cpu, sub_gpu, float_check, path + [index])
[2022-08-30T18:29:43.748Z]     
[2022-08-30T18:29:43.748Z]                 index = index + 1
[2022-08-30T18:29:43.748Z]         elif (t is dict):
[2022-08-30T18:29:43.748Z]             # The order of key/values is not guaranteed in python dicts, nor are they guaranteed by Spark
[2022-08-30T18:29:43.748Z]             # so sort the items to do our best with ignoring the order of dicts
[2022-08-30T18:29:43.748Z]             cpu_items = list(cpu.items()).sort(key=_RowCmp)
[2022-08-30T18:29:43.748Z]             gpu_items = list(gpu.items()).sort(key=_RowCmp)
[2022-08-30T18:29:43.748Z]             _assert_equal(cpu_items, gpu_items, float_check, path + ["map"])
[2022-08-30T18:29:43.748Z]         elif (t is int):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU int values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif (t is float):
[2022-08-30T18:29:43.748Z]             if (math.isnan(cpu)):
[2022-08-30T18:29:43.748Z]                 assert math.isnan(gpu), "GPU and CPU float values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]             else:
[2022-08-30T18:29:43.748Z]                 assert float_check(cpu, gpu), "GPU and CPU float values are different {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, str):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU string values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, datetime):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU timestamp values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, date):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU date values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, bool):
[2022-08-30T18:29:43.748Z]             assert cpu == gpu, "GPU and CPU boolean values are different at {}".format(path)
[2022-08-30T18:29:43.748Z]         elif isinstance(cpu, Decimal):
[2022-08-30T18:29:43.748Z] >           assert cpu == gpu, "GPU and CPU decimal values are different at {}".format(path)
[2022-08-30T18:29:43.748Z] E           AssertionError: GPU and CPU decimal values are different at [47, 'sum(b)']
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] ../../src/main/python/asserts.py:92: AssertionError
[2022-08-30T18:29:43.748Z] ----------------------------- Captured stdout call -----------------------------
[2022-08-30T18:29:43.748Z] ### CPU RUN ###
[2022-08-30T18:29:43.748Z] ### GPU RUN ###
[2022-08-30T18:29:43.748Z] ### COLLECT: GPU TOOK 0.23318719863891602 CPU TOOK 0.24316978454589844 ###
[2022-08-30T18:29:43.748Z] CPU OUTPUT: [Row(a=-32619, sum(b)=None), Row(a=-30928, sum(b)=Decimal('43613799014647028251830949477242969843')), Row(a=-28447, sum(b)=Decimal('-83725569466846329443175710034175061750')), Row(a=-27704, sum(b)=Decimal('83673968153546004357978039265761081121')), Row(a=-24513, sum(b)=None), Row(a=-23103, sum(b)=Decimal('-73881712366500820608114571040007436072')), Row(a=-20995, sum(b)=Decimal('-43285081513864731207507460740810091460')), Row(a=-20339, sum(b)=Decimal('78026661452699550461353941661132082832')), Row(a=-19569, sum(b)=Decimal('37944782979790961693390352992571616855')), Row(a=-17649, sum(b)=Decimal('-63811224080804710763061284537997550196')), Row(a=-15700, sum(b)=Decimal('-17835153556073193480854495404826110615')), Row(a=-14091, sum(b)=Decimal('-77578447058772832613753473514625789601')), Row(a=-13506, sum(b)=Decimal('-11865363805792297674977806356804951307')), Row(a=-7945, sum(b)=Decimal('21642604475137131231503773004203997695')), Row(a=-6134, sum(b)=Decimal('65182502818993305090352473765013963143')), Row(a=-5967, sum(b)=None), Row(a=-4137, sum(b)=Decimal('33111966566756745404264093656831713319')), Row(a=-3709, sum(b)=Decimal('21867479065020489288063726466395621356')), Row(a=-1901, sum(b)=Decimal('4729335124079427289918337653286636193')), Row(a=-1035, sum(b)=Decimal('-12777901277508984890515648228508830240')), Row(a=-913, sum(b)=Decimal('-8163431550632523582283561564508231025')), Row(a=1, sum(b)=Decimal('61905973175579656937274349688102686361')), Row(a=66, sum(b)=None), Row(a=1318, sum(b)=Decimal('63636197454468312157141065752120354598')), Row(a=1375, sum(b)=Decimal('-37106304908383885223095895931328039293')), Row(a=3359, sum(b)=Decimal('-59682615996904006072299884545466058266')), Row(a=4173, sum(b)=Decimal('753217583143941892444266253113610777')), Row(a=4935, sum(b)=Decimal('-3525048965644564623348132531719032994')), Row(a=5385, sum(b)=Decimal('-752015698258531592517770267898834856')), Row(a=6987, sum(b)=Decimal('-55449721492552717064150016930487718249')), Row(a=8676, sum(b)=Decimal('34114832387670067691338977572134299574')), Row(a=9182, sum(b)=Decimal('54628090452150896307406820551619391806')), Row(a=9857, sum(b)=Decimal('2688551847197926787241363712285667881')), Row(a=10511, sum(b)=Decimal('23829901252007883888494414498035548199')), Row(a=10846, sum(b)=None), Row(a=10896, sum(b)=Decimal('-36635001565727457654572234976512213543')), Row(a=11176, sum(b)=Decimal('-49903919604467173684789641371226472273')), Row(a=14072, sum(b)=Decimal('-75741626831426962049074365723075317754')), Row(a=17681, sum(b)=Decimal('25540633446776025449687473521701111467')), Row(a=18519, sum(b)=Decimal('-43213708809092373414707178568975625170')), Row(a=19506, sum(b)=Decimal('-11531723562438250311691614317433377622')), Row(a=22206, sum(b)=None), Row(a=22357, sum(b)=Decimal('-74475523813202712337620849352375571012')), Row(a=25948, sum(b)=Decimal('-70518383647729405524064484412608892422')), Row(a=26288, sum(b)=Decimal('-44675581455047983491819071764736994108')), Row(a=30923, sum(b)=Decimal('-96375396703039112292881835675701043809')), Row(a=31363, sum(b)=None), Row(a=32767, sum(b)=Decimal('-10069950567342662336581895246262384447'))]
[2022-08-30T18:29:43.748Z] GPU OUTPUT: [Row(a=-32619, sum(b)=None), Row(a=-30928, sum(b)=Decimal('43613799014647028251830949477242969843')), Row(a=-28447, sum(b)=Decimal('-83725569466846329443175710034175061750')), Row(a=-27704, sum(b)=Decimal('83673968153546004357978039265761081121')), Row(a=-24513, sum(b)=None), Row(a=-23103, sum(b)=Decimal('-73881712366500820608114571040007436072')), Row(a=-20995, sum(b)=Decimal('-43285081513864731207507460740810091460')), Row(a=-20339, sum(b)=Decimal('78026661452699550461353941661132082832')), Row(a=-19569, sum(b)=Decimal('37944782979790961693390352992571616855')), Row(a=-17649, sum(b)=Decimal('-63811224080804710763061284537997550196')), Row(a=-15700, sum(b)=Decimal('-17835153556073193480854495404826110615')), Row(a=-14091, sum(b)=Decimal('-77578447058772832613753473514625789601')), Row(a=-13506, sum(b)=Decimal('-11865363805792297674977806356804951307')), Row(a=-7945, sum(b)=Decimal('21642604475137131231503773004203997695')), Row(a=-6134, sum(b)=Decimal('65182502818993305090352473765013963143')), Row(a=-5967, sum(b)=None), Row(a=-4137, sum(b)=Decimal('33111966566756745404264093656831713319')), Row(a=-3709, sum(b)=Decimal('21867479065020489288063726466395621356')), Row(a=-1901, sum(b)=Decimal('4729335124079427289918337653286636193')), Row(a=-1035, sum(b)=Decimal('-12777901277508984890515648228508830240')), Row(a=-913, sum(b)=Decimal('-8163431550632523582283561564508231025')), Row(a=1, sum(b)=Decimal('61905973175579656937274349688102686361')), Row(a=66, sum(b)=None), Row(a=1318, sum(b)=Decimal('63636197454468312157141065752120354598')), Row(a=1375, sum(b)=Decimal('-37106304908383885223095895931328039293')), Row(a=3359, sum(b)=Decimal('-59682615996904006072299884545466058266')), Row(a=4173, sum(b)=Decimal('753217583143941892444266253113610777')), Row(a=4935, sum(b)=Decimal('-3525048965644564623348132531719032994')), Row(a=5385, sum(b)=Decimal('-752015698258531592517770267898834856')), Row(a=6987, sum(b)=Decimal('-55449721492552717064150016930487718249')), Row(a=8676, sum(b)=Decimal('34114832387670067691338977572134299574')), Row(a=9182, sum(b)=Decimal('54628090452150896307406820551619391806')), Row(a=9857, sum(b)=Decimal('2688551847197926787241363712285667881')), Row(a=10511, sum(b)=Decimal('23829901252007883888494414498035548199')), Row(a=10846, sum(b)=None), Row(a=10896, sum(b)=Decimal('-36635001565727457654572234976512213543')), Row(a=11176, sum(b)=Decimal('-49903919604467173684789641371226472273')), Row(a=14072, sum(b)=Decimal('-75741626831426962049074365723075317754')), Row(a=17681, sum(b)=Decimal('25540633446776025449687473521701111467')), Row(a=18519, sum(b)=Decimal('-43213708809092373414707178568975625170')), Row(a=19506, sum(b)=Decimal('-11531723562438250311691614317433377622')), Row(a=22206, sum(b)=None), Row(a=22357, sum(b)=Decimal('-74475523813202712337620849352375571012')), Row(a=25948, sum(b)=Decimal('-70518383647729405524064484412608892422')), Row(a=26288, sum(b)=Decimal('-44675581455047983491819071764736994108')), Row(a=30923, sum(b)=Decimal('-96375396703039112292881835675701043809')), Row(a=31363, sum(b)=None), Row(a=32767, sum(b)=None)]
[2022-08-30T18:29:43.748Z] =============================== warnings summary ===============================
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z]   /var/lib/jenkins/spark/spark-3.1.2-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/mllib/linalg/__init__.py:313: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z]   /var/lib/jenkins/spark/spark-3.1.2-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/ml/linalg/__init__.py:273: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z]   /var/lib/jenkins/spark/spark-3.1.2-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/mllib/linalg/__init__.py:596: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z] src/main/python/cache_test.py::test_cache_udt
[2022-08-30T18:29:43.748Z]   /var/lib/jenkins/spark/spark-3.1.2-bin-hadoop3.2/python/lib/pyspark.zip/pyspark/ml/linalg/__init__.py:558: DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
[2022-08-30T18:29:43.748Z] 
[2022-08-30T18:29:43.748Z] -- Docs: https://docs.pytest.org/en/stable/warnings.html
[2022-08-30T18:29:43.748Z] - generated xml file: /raid1/jenkins/workspace/rapids_it-UCX-egx06-standalone/integration_tests/target/run_dir/TEST-pytest-1661864734029257772.xml -
[2022-08-30T18:29:43.748Z] =========================== short test summary info ============================
[2022-08-30T18:29:43.748Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_grpby_sum_full_decimal[{'spark.rapids.sql.variableFloatAgg.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.castStringToFloat.enabled': 'true'}-[('a', RepeatSeq(Short)), ('b', Decimal(38,-10)), ('c', Decimal(38,-10))]][IGNORE_ORDER, INCOMPAT, APPROXIMATE_FLOAT]
[2022-08-30T18:29:43.749Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_grpby_sum_full_decimal[{'spark.rapids.sql.variableFloatAgg.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hashAgg.replaceMode': 'final'}-[('a', RepeatSeq(Short)), ('b', Decimal(38,0)), ('c', Decimal(38,0))]][IGNORE_ORDER, INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(HashAggregateExec,AggregateExpression,UnscaledValue,MakeDecimal,AttributeReference,Alias,Sum,Count,Max,Min,Average,Cast,StddevPop,StddevSamp,VariancePop,VarianceSamp,KnownFloatingPointNormalized,NormalizeNaNAndZero,GreaterThan,Literal,If,EqualTo,First,SortAggregateExec,Coalesce,IsNull,EqualNullSafe,PivotFirst,GetArrayItem,ShuffleExchangeExec,HashPartitioning)]
[2022-08-30T18:29:43.749Z] FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_grpby_sum_full_decimal[{'spark.rapids.sql.variableFloatAgg.enabled': 'true', 'spark.rapids.sql.hasNans': 'false', 'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hashAgg.replaceMode': 'partial'}-[('a', RepeatSeq(Short)), ('b', Decimal(38,0)), ('c', Decimal(38,0))]][IGNORE_ORDER, INCOMPAT, APPROXIMATE_FLOAT, ALLOW_NON_GPU(HashAggregateExec,AggregateExpression,UnscaledValue,MakeDecimal,AttributeReference,Alias,Sum,Count,Max,Min,Average,Cast,StddevPop,StddevSamp,VariancePop,VarianceSamp,KnownFloatingPointNormalized,NormalizeNaNAndZero,GreaterThan,Literal,If,EqualTo,First,SortAggregateExec,Coalesce,IsNull,EqualNullSafe,PivotFirst,GetArrayItem,ShuffleExchangeExec,HashPartitioning)]
[2022-08-30T18:29:43.749Z] = 3 failed, 13947 passed, 662 skipped, 464 xfailed, 196 xpassed, 12 warnings in 19444.75s (5:24:04) =
```

--- a/docs/compatibility.md
+++ b/docs/compatibility.md
@@ -111,15 +111,6 @@ a few operations that we cannot support to the same degree as Spark can on the C
 
 ### Decimal Sum Aggregation
 
-A number of fixes for overflow detection went into Spark 3.1.0. Please see
-[SPARK-28067](https://issues.apache.org/jira/browse/SPARK-28067) and
-[SPARK-32018](https://issues.apache.org/jira/browse/SPARK-32018) for more detailed information.
-Some of these fixes we were able to back port, but some of them require Spark 3.1.0 or above to
-fully be able to detect overflow in all cases. As such on versions of Spark older than 3.1.0 for
-large decimal values there is the possibility of data corruption in some corner cases. 
-This is true for both the CPU and GPU implementations, but there are fewer of these cases for the 
-GPU. If this concerns you, you should upgrade to Spark 3.1.0 or above. 
-
 When Apache Spark does a sum aggregation on decimal values it will store the result in a value
 with a precision that is the input precision + 10, but with a maximum precision of 38.
 For an input precision of 9 and above, Spark will do the aggregations as a Java `BigDecimal`
@@ -131,8 +122,13 @@ longer detected. Even then all the values would need to be either the largest or
 possible to be stored in the type for the overflow to cause data corruption.
 
 For the RAPIDS Accelerator we don't have direct access to unlimited precision for our calculations
-like the CPU does. For input values with a precision of 8 and below we follow Spark and process the
-data the same way, as a 64-bit value. For larger values we will do extra calculations looking at the
+like the CPU does, and the aggregations are processed in batches within each task. Therefore it is
+possible for the GPU to detect an intermediate overflow after aggregating a batch, e.g.: a sum
+aggregation on positive and negative values, where the accumulating sum value temporarily
+overflows but returns within bounds before the final cast back into a decimal with precision 38.
+
+For input values with a precision of 8 and below we follow Spark and process the data the
+same way, as a 64-bit value. For larger values we will do extra calculations looking at the
 higher order digits to be able to detect overflow in all cases. But because of this you may see
 some performance differences depending on the input precision used. The differences will show up
 when going from an input precision of 8 to 9 and again when going from an input precision of 28 to 29.

--- a/integration_tests/src/main/python/data_gen.py
+++ b/integration_tests/src/main/python/data_gen.py
@@ -215,19 +215,22 @@ class IntegerGen(DataGen):
 
 class DecimalGen(DataGen):
     """Generate Decimals, with some built in corner cases."""
-    def __init__(self, precision=None, scale=None, nullable=True, special_cases=None):
+    def __init__(self, precision=None, scale=None, nullable=True, special_cases=None, avoid_positive_values=False):
         if precision is None:
             #Maximum number of decimal digits a Long can represent is 18
             precision = 18
             scale = 0
         DECIMAL_MIN = Decimal('-' + ('9' * precision) + 'e' + str(-scale))
         DECIMAL_MAX = Decimal(('9'* precision) + 'e' + str(-scale))
-        if (special_cases is None):
-            special_cases = [DECIMAL_MIN, DECIMAL_MAX, Decimal('0')]
+        if special_cases is None:
+            special_cases = [DECIMAL_MIN, Decimal('0')]
+            if not avoid_positive_values:
+                special_cases.append(DECIMAL_MAX)
         super().__init__(DecimalType(precision, scale), nullable=nullable, special_cases=special_cases)
         self.scale = scale
         self.precision = precision
-        pattern = "-?[0-9]{1,"+ str(precision) + "}e" + str(-scale)
+        negative_pattern = "-" if avoid_positive_values else "-?"
+        pattern = negative_pattern + "[0-9]{1,"+ str(precision) + "}e" + str(-scale)
         self.base_strs = sre_yield.AllStrings(pattern, flags=0, charset=sre_yield.CHARSET, max_count=_MAX_CHOICES)
 
     def __repr__(self):

--- a/integration_tests/src/main/python/hash_aggregate_test.py
+++ b/integration_tests/src/main/python/hash_aggregate_test.py
@@ -218,9 +218,7 @@ _no_overflow_ansi_gens = [
 
 _decimal_gen_36_5 = DecimalGen(precision=36, scale=5)
 _decimal_gen_36_neg5 = DecimalGen(precision=36, scale=-5)
-_decimal_gen_38_0 = DecimalGen(precision=38, scale=0)
 _decimal_gen_38_10 = DecimalGen(precision=38, scale=10)
-_decimal_gen_38_neg10 = DecimalGen(precision=38, scale=-10)
 
 
 def get_params(init_list, marked_params=[]):
@@ -303,15 +301,22 @@ _grpkey_short_very_big_neg_scale_decimals = [
     ('b', _decimal_gen_36_neg5),
     ('c', _decimal_gen_36_neg5)]
 
-_grpkey_short_full_decimals = [
-    ('a', RepeatSeqGen(short_gen, length=50)),
-    ('b', _decimal_gen_38_0),
-    ('c', _decimal_gen_38_0)]
+# Only use negative values to avoid the potential to hover around an overflow
+# as values are added and subtracted during the sum. Non-deterministic ordering
+# of values from shuffle cannot guarantee overflow calculation is predictable
+# when the sum can move in both directions as new partitions are aggregated.
+_decimal_gen_sum_38_0 = DecimalGen(precision=38, scale=0, avoid_positive_values=True)
+_decimal_gen_sum_38_neg10 = DecimalGen(precision=38, scale=-10, avoid_positive_values=True)
 
-_grpkey_short_full_neg_scale_decimals = [
+_grpkey_short_sum_full_decimals = [
     ('a', RepeatSeqGen(short_gen, length=50)),
-    ('b', _decimal_gen_38_neg10),
-    ('c', _decimal_gen_38_neg10)]
+    ('b', _decimal_gen_sum_38_0),
+    ('c', _decimal_gen_sum_38_0)]
+
+_grpkey_short_sum_full_neg_scale_decimals = [
+    ('a', RepeatSeqGen(short_gen, length=50)),
+    ('b', _decimal_gen_sum_38_neg10),
+    ('c', _decimal_gen_sum_38_neg10)]
 
 
 _init_list_no_nans_with_decimal = _init_list_no_nans + [
@@ -327,9 +332,6 @@ _init_list_with_nans_and_no_nans_with_decimalbig = _init_list_with_nans_and_no_n
     _grpkey_short_big_decimals, _grpkey_short_very_big_decimals, 
     _grpkey_short_very_big_neg_scale_decimals]
 
-
-_init_list_full_decimal = [_grpkey_short_full_decimals, 
-    _grpkey_short_full_neg_scale_decimals]
 
 #Any smaller precision takes way too long to process on the CPU
 # or results in using too much memory on the GPU
@@ -386,7 +388,7 @@ def test_hash_grpby_sum(data_gen, conf):
 @approximate_float
 @ignore_order
 @incompat
-@pytest.mark.parametrize('data_gen', _init_list_full_decimal, ids=idfn)
+@pytest.mark.parametrize('data_gen', [_grpkey_short_sum_full_decimals, _grpkey_short_sum_full_neg_scale_decimals], ids=idfn)
 @pytest.mark.parametrize('conf', get_params(_confs, params_markers_for_confs), ids=idfn)
 def test_hash_grpby_sum_full_decimal(data_gen, conf):
     assert_gpu_and_cpu_are_equal_collect(
@@ -418,7 +420,7 @@ def test_hash_reduction_sum_full_decimal(data_gen, conf):
 @ignore_order
 @incompat
 @pytest.mark.parametrize('data_gen', _init_list_with_nans_and_no_nans + [_grpkey_short_mid_decimals, 
-    _grpkey_short_big_decimals, _grpkey_short_very_big_decimals, _grpkey_short_full_decimals], ids=idfn)
+    _grpkey_short_big_decimals, _grpkey_short_very_big_decimals, _grpkey_short_sum_full_decimals], ids=idfn)
 @pytest.mark.parametrize('conf', get_params(_confs, params_markers_for_confs), ids=idfn)
 def test_hash_grpby_avg(data_gen, conf):
     assert_gpu_and_cpu_are_equal_collect(

